tests/test_visitors/test_ast/test_annotations/test_argument_annotations.py

```diff
@@ -68,6 +90,85 @@ def function(empty_arg: Literal[None]): ...
 def function(empty_arg: Union[Literal[None], Optional[int]]): ...
 """
 
+wrong_nested_literal_annotation = """
+def function(arg: Literal[Literal[1, 2], 3]): ...
+"""
+
+wrong_nested_union_annotation = """
+def function(arg: Union[Union[int, str], float]): ...
+"""
+
+wrong_nested_annotated_annotation = """
+def function(arg: Annotated[Annotated[int, str], float]): ...
+"""
+
+wrong_deep_nested_literal_annotation = """
+def function(arg: Literal[Literal[Literal[1]]]): ...
+"""
+
+wrong_deep_nested_union_annotation = """
+def function(arg: Union[Union[Union[int]]]): ...
+"""
+
+wrong_deep_nested_annotated_annotation = """
+def function(arg: Annotated[Annotated[Annotated[int]]]): ...
+"""
+
+wrong_nested_prefixed_literal_annotation = """
+def function(arg: typing.Literal[typing.Literal[1]]): ...
+"""
+
+wrong_deep_nested_prefixed_literal_annotation = """
+def function(arg: typing.Literal[typing.Literal[typing.Literal[1]]]): ...
+"""
+
+
+@pytest.mark.parametrize('code', [
+    wrong_nested_literal_annotation,
+    wrong_nested_union_annotation,
+    wrong_nested_annotated_annotation,
+    wrong_nested_prefixed_literal_annotation,
+])
+def test_forbidden_nested_annotations(
+    assert_errors,
+    parse_ast_tree,
+    code,
+    default_options,
+    mode,
+):
+    """Ensures that using nested ``Literal`` and ``Union`` is forbidden."""
+    tree = parse_ast_tree(mode(code))
+
+    visitor = WrongNestedAnnotationVisitor(default_options, tree=tree)
+    visitor.run()
+
+    assert_errors(visitor, [NestedAnnotationsViolation])
+
+
+@pytest.mark.parametrize('code', [
+    wrong_deep_nested_literal_annotation,
+    wrong_deep_nested_union_annotation,
+    wrong_deep_nested_annotated_annotation,
+    wrong_deep_nested_prefixed_literal_annotation,
+])
+def test_forbidden_deep_nested_annotations(
```

Comment:
What about mixing `Union[]` and `typing.Union[]`? I know that it is a rare case, but we must handle it correctly.

Reply:
Done!